test/shared/src/main/scala/zio/test/TestAspect.scala

```diff
@@ -441,50 +480,82 @@ object TestAspect extends TimeoutVariants {
   /**
    * An aspect that retries failed tests according to a schedule.
    */
-  def retry[R0, E0, S0](
+  def retry[R0 <: TestClock with TestConsole with TestRandom with TestSystem, E0, S0](
     schedule: Schedule[R0, TestFailure[E0], S0]
   ): TestAspect[Nothing, R0, Nothing, E0, Nothing, Any] =
-    new TestAspect.PerTest[Nothing, R0, Nothing, E0, Nothing, Any] {
-      def perTest[R >: Nothing <: R0, E >: Nothing <: E0, S >: Nothing <: Any](
-        test: ZIO[R, TestFailure[E], TestSuccess[S]]
-      ): ZIO[R, TestFailure[E], TestSuccess[S]] =
-        test.retry(schedule)
-    }
+    restoreTestEnvironment >>>
+      new TestAspect.PerTest[Nothing, R0, Nothing, E0, Nothing, Any] {
+        def perTest[R >: Nothing <: R0, E >: Nothing <: E0, S >: Nothing <: Any](
+          test: ZIO[R, TestFailure[E], TestSuccess[S]]
+        ): ZIO[R, TestFailure[E], TestSuccess[S]] =
+          test.retry(schedule)
+      }
 
   /**
-   * An aspect that executes the members of a suite sequentially.
+   * An aspect that restores the [[TestClock]]'s state to its starting state after the test is run.
+   * Note that this is only useful when repeating tests.
    */
-  val sequential: TestAspectPoly = executionStrategy(ExecutionStrategy.Sequential)
+  def restoreTestClock: TestAspect[Nothing, TestClock, Nothing, Any, Nothing, Any] =
+    new TestAspect.PerTest[Nothing, TestClock, Nothing, Any, Nothing, Any] {
+      def perTest[R >: Nothing <: TestClock, E >: Nothing <: Any, S >: Nothing <: Any](
+        test: ZIO[R, TestFailure[E], TestSuccess[S]]
+      ): ZIO[R, TestFailure[E], TestSuccess[S]] =
+        TestClock.save.flatMap(test <* _)
+    }
 
   /**
-   * An aspect that applies the specified aspect on Scala 2.
+   * An aspect that restores the [[TestConsole]]'s state to its starting state after the test is run.
+   * Note that this is only useful when repeating tests.
    */
-  def scala2[LowerR, UpperR, LowerE, UpperE, LowerS, UpperS](
-    that: TestAspect[LowerR, UpperR, LowerE, UpperE, LowerS, UpperS]
-  ): TestAspect[LowerR, UpperR, LowerE, UpperE, LowerS, UpperS] =
-    if (TestVersion.isScala2) that else identity
+  def restoreTestConsole: TestAspect[Nothing, TestConsole, Nothing, Any, Nothing, Any] =
```

Comment:
Can we write a `restore` test aspect that works for any `R` that extends `Restorable` and then implement all these methods in terms of that?

Reply:
@adamgfraser  this is a great idea but the short answer is "not successfully to date" :) Can you please take a look at my attempt so far and see if you can point me in the right direction?